### ISSUE TITLE
Remove Warning about no user-provided default constructor

### ISSUE THREAD

**Issue by gajanak** [msgpack/msgpack-c#711]
Hello,

I experiment with the x3_parse and get a warning:

    note: 'union msgpack::v2::detail::<lambda(auto:18&)> [blahblah] ::<unnamed>' has no user-provided default constructor
     union { uint32_t i; float f; } mem;
                ^
    note: and the implicitly-defined constructor does not initialize 'uint32_t msgpack::v2::detail::<lambda(auto:18&)> [blahblah]  ::<unnamed union>::i'

Looking on the Code, I added the following change. I don't understand the whole code, so perhaps someone with more knowledge can me explain my fault. With this code, its working.
I thought its standard syntax. (Using gcc 8.1  with c++17 )

```diff
diff --git a/include/msgpack/v2/x3_parse.hpp b/include/msgpack/v2/x3_parse.hpp
index 09612f5..9b033b2 100644
--- a/include/msgpack/v2/x3_parse.hpp
+++ b/include/msgpack/v2/x3_parse.hpp
@@ -297,8 +297,7 @@ const auto mp_object_def =
         (
             [](auto& ctx){
                 auto& app_specific = x3::get<tag_app_specific>(ctx).get();
-                union { uint32_t i; float f; } mem;
-                mem.i = _attr(ctx);
+                union { uint32_t i; float f; } mem = {i:_attr(ctx) };
                 app_specific.vis.visit_float32(mem.f);
             }
         )
@@ -309,8 +308,7 @@ const auto mp_object_def =
         (
             [](auto& ctx){
                 auto& app_specific = x3::get<tag_app_specific>(ctx).get();
-                union { uint64_t i; double f; } mem;
-                mem.i = _attr(ctx);
+                union { uint64_t i; double f; } mem= {i:_attr(ctx) };
 #if defined(TARGET_OS_IPHONE)
                 // ok
 #elif defined(__arm__) && !(__ARM_EABI__) // arm-oabi
```

Is this a possible solution ?  For me it looks better. ;)

**Comment by redboltz** [comment]
I got `warning: use of GNU old-style field designator extension [-Wgnu-designator]`.

https://wandbox.org/permlink/BNFRrgGYw6cr8YpM

If I use the following notation, I got `warning: designated initializers are a C99 feature [-Wc99-extensions]`

```cpp
    union { uint64_t i; double f; } mem= { .i = 1 };
```

https://wandbox.org/permlink/Znk6Uwfx0jyXo597

**Comment by gajanak** [comment]
Hm, looking at https://en.cppreference.com/w/cpp/language/aggregate_initialization
Ok, designated initalizer is c++20 ...
Perhaps this is more compatible:

```cpp
int main() {
    union { uint64_t i; double f; } mem= { 1 }; // OK, first member of the union
}
```
looks warnign free in gcc & clang  old&new one.

**Comment by gajanak** [comment]
Other solution for type-punning would be perhaps `return *reinterpret_cast<float*>( &val );` 
But I personaly prefer the union, too. ;)

**Comment by redboltz** [comment]
Fixed by #720
